File: gatling-recorder/src/main/java/com/excilys/ebi/gatling/recorder/http/handler/AbstractBrowserRequestHandler.java
/**
 * Copyright 2011 eBusiness Information, Groupe Excilys (www.excilys.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.excilys.ebi.gatling.recorder.http.handler;

import static com.excilys.ebi.gatling.recorder.http.event.RecorderEventBus.getEventBus;

import org.jboss.netty.channel.ChannelFuture;
import org.jboss.netty.channel.ChannelFutureListener;
import org.jboss.netty.channel.ChannelHandlerContext;
import org.jboss.netty.channel.ExceptionEvent;
import org.jboss.netty.channel.MessageEvent;
import org.jboss.netty.channel.SimpleChannelHandler;
import org.jboss.netty.handler.codec.http.HttpRequest;

import com.excilys.ebi.gatling.recorder.http.event.MessageReceivedEvent;
import com.excilys.ebi.gatling.recorder.http.event.RequestReceivedEvent;

public abstract class AbstractBrowserRequestHandler extends SimpleChannelHandler {

	protected final String outgoingProxyHost;
	protected final int outgoingProxyPort;

	public AbstractBrowserRequestHandler(String outgoingProxyHost, int outgoingProxyPort) {
		this.outgoingProxyHost = outgoingProxyHost;
		this.outgoingProxyPort = outgoingProxyPort;
	}

	@Override
	public void messageReceived(ChannelHandlerContext ctx, MessageEvent event) throws Exception {

		getEventBus().post(new MessageReceivedEvent(ctx.getChannel()));

		HttpRequest request = HttpRequest.class.cast(event.getMessage());
		
		// remove Proxy-Connection header if it's not significant
		if (outgoingProxyHost == null)
			request.removeHeader("Proxy-Connection");

		ChannelFuture future = connectToServerOnBrowserRequestReceived(ctx, request);
		
		getEventBus().post(new RequestReceivedEvent(request));
		
		sendRequestToServerAfterConnection(future, request);
	}

	protected abstract ChannelFuture connectToServerOnBrowserRequestReceived(ChannelHandlerContext ctx, HttpRequest request) throws Exception;

	@Override
	public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e) throws Exception {
		System.err.println("Exception caught = ");
		e.getCause().printStackTrace();

		// Properly closing
		ChannelFuture future = ctx.getChannel().getCloseFuture();
		future.addListener(new ChannelFutureListener() {
			@Override
			public void operationComplete(ChannelFuture future) throws Exception {
				future.getChannel().close();
			}
		});
		ctx.sendUpstream(e);
	}

	private void sendRequestToServerAfterConnection(ChannelFuture future, final HttpRequest request) {
		if (future != null)
			future.addListener(new ChannelFutureListener() {
				@Override
				public void operationComplete(ChannelFuture future) throws Exception {
					future.getChannel().write(request);
				}
			});
	}
}

File: gatling-recorder/src/main/scala/com/excilys/ebi/gatling/recorder/http/handler/AbstractBrowserRequestHandler.scala
/**
 * Copyright 2011-2012 eBusiness Information, Groupe Excilys (www.excilys.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 		http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.excilys.ebi.gatling.recorder.http.handler

import java.net.URI

import scala.collection.JavaConversions.asScalaBuffer

import org.jboss.netty.channel.{ ChannelFuture, ChannelFutureListener, ChannelHandlerContext, ExceptionEvent, MessageEvent, SimpleChannelHandler }
import org.jboss.netty.handler.codec.http.{ DefaultHttpRequest, HttpRequest }

import com.excilys.ebi.gatling.http.Headers
import com.excilys.ebi.gatling.recorder.config.ProxyConfig
import com.excilys.ebi.gatling.recorder.controller.RecorderController
import com.ning.http.util.Base64

import grizzled.slf4j.Logging

abstract class AbstractBrowserRequestHandler(controller: RecorderController, proxyConfig: ProxyConfig) extends SimpleChannelHandler with Logging {

	override def messageReceived(ctx: ChannelHandlerContext, event: MessageEvent) {

		event.getMessage match {
			case request: HttpRequest =>
				proxyConfig.host.map { _ =>
					for {
						username <- proxyConfig.username
						password <- proxyConfig.password
					} {
						val proxyAuth = "Basic " + Base64.encode((username + ":" + password).getBytes)
						request.setHeader(Headers.Names.PROXY_AUTHORIZATION, proxyAuth)
					}
				}.getOrElse(request.removeHeader("Proxy-Connection")) // remove Proxy-Connection header if it's not significant

				val future = connectToServerOnBrowserRequestReceived(ctx, request)

				controller.receiveRequest(request)

				sendRequestToServerAfterConnection(future, request);

			case unknown => warn("Received unknown message: " + unknown)
		}
	}

	def connectToServerOnBrowserRequestReceived(ctx: ChannelHandlerContext, request: HttpRequest): ChannelFuture

	override def exceptionCaught(ctx: ChannelHandlerContext, e: ExceptionEvent) {
		error("Exception caught", e.getCause)

		// Properly closing
		val future = ctx.getChannel.getCloseFuture
		future.addListener(new ChannelFutureListener {
			def operationComplete(future: ChannelFuture) = future.getChannel.close
		})
		ctx.sendUpstream(e)
	}

	private def sendRequestToServerAfterConnection(future: ChannelFuture, request: HttpRequest) {

		Option(future).map { future =>
			future.addListener(new ChannelFutureListener {
				def operationComplete(future: ChannelFuture) = future.getChannel.write(buildRequestWithRelativeURI(request))
			})
		}
	}

	private def buildRequestWithRelativeURI(request: HttpRequest) = {
		val uri = new URI(request.getUri)
		val newUri = new URI(null, null, null, -1, uri.getPath, uri.getQuery, uri.getFragment).toString
		val newRequest = new DefaultHttpRequest(request.getProtocolVersion, request.getMethod, newUri)
		newRequest.setChunked(request.isChunked)
		newRequest.setContent(request.getContent)
		for (header <- request.getHeaders) newRequest.addHeader(header.getKey, header.getValue)
		newRequest
	}
}
